---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# ScrambledTreeBuilder

<!-- badges: start -->
<!-- badges: end -->

Produce, plot and use phylogenetic trees to study genome scrambling.

## Installation

As of today there is no release. Install the development version of _ScrambledTreeBuilder_ from [GitHub](https://github.com/brennern/ScrambledTreeBuilder/) with:
      
```
# install.packages("devtools")
devtools::install_github("brennern/ScrambledTreeBuilder", repos=BiocManager::repositories()) |> suppressMessages()
```

## Usage

The input of _ScrambledTreeBuilder_ is YAML files produced by the [oist/plessy_nf_GenomicBreaks](https://github.com/oist/plessy_nf_GenomicBreaks) pipeline from all-versus-all pairwise genome alignments computed with the (nf-core/pairgenomealign)[https://nf-co.re/pairgenomealign] pipeline.  This package includes a few YAML files in its example data, and example objects derived from these files to compute and plot trees. Here is an example of tree produced by _ScrambledTreeBuilder_. See the [Get started](articles/ScrambledTreeBuilder.html) vignette for details.

```{r SingleValueTree}
library(ScrambledTreeBuilder)
visualizeTree(Halo_Tree, "Scrambling_index") +
  Halo_FocalClades +
  ggplot2::ggtitle(paste("Tree built with Percent difference and labelled with Scrambling index")) + 
  viridis::scale_color_viridis(name = "Scambling index")
```
