---
title: "ScrambledTreeBuilder"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{ScrambledTreeBuilder}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options:
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r}
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_knit$set(verbose = TRUE)
```

# Introduction

The _ScrambledTreeBuilder_ package consists of numerous data formatting
functions for phylogenetic tree building in the context of the (still internal)
[Scrambling in the Tree of Life](https://github.com/oist/LuscombeU_ScramblingInTheTreeOfLife)
project.

# Load Package

The _ScrambledTreeBuilder_ package ouptuts plots in _ggplot2_ format but you
need to load the _ggplot2_ package to further customize them.

```{r setup}
library(ScrambledTreeBuilder)
library(ggplot2)
```

# Example Data

This package utilizes example `.yaml` files containing summary statistics of
halobacteria genome comparison data. In regards to genome scrambling, many
studies have showcased significant genome rearrangments in such halobacteria
species due to dozens of insertion sequence families.

The YAML summary files are produced by performing an all vs. all genome
comparison between six halobacteria species using the Nextflow
[genome pairwise alignment pipeline](https://github.com/oist/plessy_pairwiseGenomeComparison)
and an [input file script](https://github.com/brennern/GenomeScrambling/blob/main/scripts/makeGBreaksInputFile.sh)
for [GenomicBreaks](https://github.com/oist/GenomicBreaks).  They are stored in
a directory called 'extdata/PairwiseComparisons'.  Here we prepare an object
called 'yamlFileData' that contains the path to the files.  The file names are
constructed by taking the name of one species taken as reference, adding a
separator (`___`), and then the name of one species being compared to the
reference.

```{r}
resultsDir <- system.file("extdata/PairwiseComparisons", package = "ScrambledTreeBuilder")
yamlFileData <- list.files(resultsDir, pattern = "*.yaml.bz2", full.names = TRUE)
names(yamlFileData) <- yamlFileData |> basename() |> sub(pat = ".yaml.bz2", rep="")
head(names(yamlFileData))
```

Next, we use the 'formatStats()' function to load the YAML files into a
single dataframe where each line is a pair of species and each column is
a statistic or a metadata about that species comparison.

```{r}
exDataFrame <- formatStats(yamlFileData)
```

Then, for statistics of interest, we build square matrices where rows and
columns indicate one species, and the cells at each intersection contain the
value for that pair.

We perform this task with the 'makeMatrix()' function.  It provide defaults for
missing values and self-comparisons.

```{r}
# Average nucleotide identity.  We will use it to cluster a tree.
(treeMatrix <- makeMatrix(exDataFrame, "percent_identity_global", 100, 50))
```

```{r}
# Strand randomisation index.
(valueMatrix <- makeMatrix(exDataFrame, "index_avg_strandRand", 1, 0.5))
```

We cluster the average nucleotide identity matrix (`treeMatrix`) to produce
a tree in _tibble_ format, using the `makeTidyTree()` funciton. Then we add
other statistics, here the strand randomisation index and the average nucleotide
identity, to this object using the `makeValueTibble` function.

```{r}
# Let's average the target-query and query-target replicate pairs.
(Tibble <- makeTidyTree(100 - (treeMatrix/2 + t(treeMatrix)/2)))
```

```{r}
(tibbleWithValue <- makeValueTibble(Tibble, valueMatrix, colname = "Strand_Randomisation_Index"))
```

```{r}
(tibbleWithMultipleValues <- makeValueTibble(tibbleWithValue, treeMatrix, colname = "Percent_Identity"))
```

We made multiple tables to show the step-by-step process, but typically one
would just keep the last table.  This can be done with pipes.

```{r}
makeTidyTree(100 - (treeMatrix/2 + t(treeMatrix)/2)) |>
  makeValueTibble(valueMatrix, colname = "Strand_Randomisation_Index") |>
  makeValueTibble(treeMatrix, colname = "Percent_Identity")
# Same result
```

## Plot trees

Let's use the `tibbleWithMultipleValues` object to plot trees.  In our case, we
have generated a tree built based on 100 minus percent identity values as a
distance, and computed average strand randomisation scores for all the nodes.
We can plot these values as labels on the tree.

```{r}
visualizeTree(tibbleWithMultipleValues, tibbleWithMultipleValues$Strand_Randomisation_Index) +
  ggplot2::ggtitle(paste("Tree built with Percent Identity and labelled with Strand Randomisation Index scores")) + 
  ggtree::geom_hilight(node = 8, fill = "lightblue1", alpha = .4, type = "gradient", gradient.direction = 'tr') +
  ggtree::geom_hilight(node = 9, fill = "pink", alpha = .4, type = "gradient", gradient.direction = 'tr')
```

```{r}
visualizeTree(tibbleWithMultipleValues, tibbleWithMultipleValues$Strand_Randomisation_Index, ynudge = 0.2) +
  ggplot2::ggtitle("Tree labeled with Strand Randomisation Index and Percent Identity (built with Percent Identity)") +
  ggplot2::scale_color_viridis_c(name = "Strand Randomisation Index", option = "cividis") +
  ggnewscale::new_scale_colour() +
  ggtree::geom_label(ggtree::aes(label=round(Percent_Identity, digits = 3), color = Percent_Identity), label.size = 0.25, size = 3, na.rm = TRUE, label.padding = ggtree::unit(0.15, "lines"), nudge_y = -0.2) +
  viridis::scale_color_viridis(option = "magma", name = "Percent Identity")
```

Of course, if you spotted an interesting sub-tree, you can plot the node IDs
to easily extract it for further analysis.

```{r}
visualizeTree(tibbleWithMultipleValues)
```

The `subTree` function can conveniently be used with R's pipe operator to cut
a sub-tree at a chosen node.

```{r}
visualizeTree(tibbleWithMultipleValues |> subTree(node = 9), "Percent_Identity")
```

```{r}
subMatrix(Tibble, valueMatrix, 9, simpl=TRUE)
```


## Plot node and leaf values

We can also plot leaf and node values in a standard scatterplot.

```{r}
plot(exDataFrame$percent_identity_global, exDataFrame$index_avg_strandRand, col = 'grey', xlab = "Percent identity", ylab="Strand randomisation index", main = "Leaf values are in grey and node values in red")
points(tibbleWithMultipleValues$Percent_Identity, tibbleWithMultipleValues$Strand_Randomisation_Index, col = 'red')

plotTwoBranches <- function(tree, node, X, Y, ...) {
  children <- tidytree::child   (tree, node)
  if(nrow(children) != 2) return(invisible())
  parent   <- tidytree::parent(tree, children$node[1])
  lines(
    c(parent[, X, drop=T], children[, X, drop=T][1]),
    c(parent[, Y, drop=T], children[, Y, drop=T][1]),
    ...
  )
  lines(
    c(parent[, X, drop=T], children[, X, drop=T][2]),
    c(parent[, Y, drop=T], children[, Y, drop=T][2]),
    ...
  )
}

plotAllBranches <- function(tree, X, Y, ... ) {
  unique(tree$parent) |> sort() |> sapply(\(node) {
    plotTwoBranches(tree, node, X, Y, ...)
  })
  return(invisible())
}

plotAllBranches(tibbleWithMultipleValues, "Percent_Identity", "Strand_Randomisation_Index", col = "red")
```
